libraries/chain/db_update.cpp

```diff
@@ -388,6 +389,16 @@ void database::clear_expired_orders()
 
          price settlement_price = pays / receives;
 
+         // Calculate fill_price with a bigger volume to reduce impacts of rounding
+         if( settlement_fill_price.base.asset_id != pays.asset_id )
+         {
+            asset tmp_pays = max_settlement_volume;
+            asset tmp_receives = tmp_pays * mia.current_feed.settlement_price;
+            tmp_receives.amount = (fc::uint128_t(tmp_receives.amount.value) *
+                            (GRAPHENE_100_PERCENT - mia.options.force_settlement_offset_percent) / GRAPHENE_100_PERCENT).to_uint64();
+            settlement_fill_price = tmp_pays / tmp_receives;
+         }
```

Comment:
IMO this block is unnecessarily complicated. max_settlement_volume is a percentage of current_supply and can potentially be very small.

The actual fill_price is equal to the settlement price multiplied by 1-offset. max_settlement_volume play no role in that calculation. The settlement offset is scaled with GRAPHENE_100_PERCENT and has therefore low precision. Unfortunately we don't have a generic means to multiply a price with a quotient...

We could use tmp_pays = asset(GRAPHENE_100_PERCENT, current_asset) and tmp_receives = asset(GRAPHENE_100_PERCENT - settlement_offset, receives.asset_id), but that could still be insufficient protection from rounding errors. OTOH if we scale too much we might get an overflow.

Or we add operators * and / to scale a price. That would be the cleanest solution, IMO.

Just thinking aloud. :-)

Reply:
IMHO `max_settlement_volume` is a good compromise in practice that usually won't be too small nor too big. Anyway, this is not a big deal, worst case is another ugly chart, so I'd like to keep this code while continuing the discussion about better implementation.

Reply:
ok, good enough for now